### PR TITLE
feat(t8s-cluster/workload-cluster): install ccm/csi via helm to inline-reuse the cloud-config secret

### PR DESCRIPTION
**Pull request by cwrau**
This replaces https://gitlab.teuto.net/4teuto/ops/k8s/t8s-engine/-/merge_requests/529 as a means of creating the `cloud-config` secret in the workload cluster

This has the advantage of a reconciling loop, including the secret and rolling the ccm if the secret changes 👍️

This is possible since [yesterday](https://github.com/kubernetes/cloud-provider-openstack/issues/2175)

This also adds pre-install jobs to remove the old resources that are taken over by HelmReleases, meaning the CCM, the CSI, the etcd-defrag and, if rolling out cilium, CNI

### CODE REVIEW

**Comment by cwrau** [comment]
@marvinWolff I'm thinking of creating a pre-install job that takes care of removing the old CCM, CSI and cloud-conf secret

That way we don't have to do this by hand, what do you think?

**Comment by marvinWolff** [comment]
> @marvinWolff I'm thinking of creating a pre-install job that takes care of removing the old CCM, CSI and cloud-conf secret
> 
> That way we don't have to do this by hand, what do you think?

I think thats a good idea.  
Can you run the pre-install job every time the ccm is updated? That's not a new install, or am i wrong?

**Review by marvinWolff — DISMISSED** [review]

**Comment by marvinWolff** [comment]
This looks good, but the MR referenced does not contain any info about this feature? :sweat_smile: 
Is it the correct one?

**Comment by cwrau** [comment]
> this feature

Do you mean the pre-install job? That's new

Do you mean this whole PR? Then I don't understand what you mean 😅

**Comment by marvinWolff** [comment]
> This has the advantage of a reconciling loop, including the secret and rolling the ccm if the secret changes +1
> 
> This is possible since [yesterday](https://github.com/kubernetes/cloud-provider-openstack/issues/2175)

No i meant the link in `yesterday`. You wrote it's possible since yesterday, therefore i thought there would be any mention of this :sweat_smile:

**Comment by cwrau** [comment]
> > This has the advantage of a reconciling loop, including the secret and rolling the ccm if the secret changes +1
> > This is possible since [yesterday](https://github.com/kubernetes/cloud-provider-openstack/issues/2175)
> 
> No i meant the link in `yesterday`. You wrote it's possible since yesterday, therefore i thought there would be any mention of this sweat_smile

Ah, the whole PR set up everything to make it usable

- they had release problems
- they now use `the version format is 2.[kubeminor].[helmpatch]`
- ...

**Comment by marvinWolff** [comment]
> > > This has the advantage of a reconciling loop, including the secret and rolling the ccm if the secret changes +1
> > > This is possible since [yesterday](https://github.com/kubernetes/cloud-provider-openstack/issues/2175)
> > 
> > 
> > No i meant the link in `yesterday`. You wrote it's possible since yesterday, therefore i thought there would be any mention of this sweat_smile
> 
> Ah, the whole PR set up everything to make it usable
> 
> * they had release problems
> * they now use `the version format is 2.[kubeminor].[helmpatch]`
> * ...

Ah ok, you meant the chart is now usable.  
I see, i thought that feature was now usable.

**Comment by cwrau** [comment]
> I think thats a good idea.

👍️ Implemented 👌

> Can you run the pre-install job every time the ccm is updated? That's not a new install, or am i wrong?

We don't need to run this multiple times, as after the first time everything is already cleaned up 👍️

File: charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      targetPath: cloudConfigContents
+      valuesKey: cloud.conf
+    - <<: *valuesFrom
+      targetPath: cloudConfig.metadata.hash
+  values:
+    tolerations:
+      - key: node.cloudprovider.kubernetes.io/uninitialized
+        value: "true"
+        effect: NoSchedule
+      - key: node-role.kubernetes.io/master
```
[the remark this answers is not in the document]
Comment (cwrau) [reply charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml:40]: I just copied this, you're right, the `master` taint is not used anymore

File: charts/t8s-cluster/templates/workload-cluster/pre-install/_uninstall-job.yaml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      serviceAccountName: {{ $name }}-uninstall
+      enableServiceLinks: false
+      containers:
+        - name: uninstall
+          image: {{ include "common.images.image" (dict "imageRoot" .Values.global.kubectl.image "global" .Values.global) }}
+          {{- if .Values.global.kubectl.image.digest }}
+          imagePullPolicy: IfNotPresent
+          {{- else }}
+          imagePullPolicy: Always
+          {{- end }}
```
[the remark this answers is not in the document]
Comment (cwrau) [reply charts/t8s-cluster/templates/workload-cluster/pre-install/_uninstall-job.yaml:39]: We can only guarantee the up-to-dateness of an image when it's using a digest, as tags can be updated.

Therefore to guarantee the up-to-dateness of an image without a digest we need to pull it every time

File: charts/t8s-cluster/templates/workload-cluster/cinder-csi-plugin.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      name: {{ .Release.Name }}-kubeconfig
+  install:
+    remediation:
+      retries: -1
+  storageNamespace: kube-system
+  targetNamespace: kube-system
+  releaseName: csi
+  values:
+    storageClass:
+      enabled: false
```
[the remark this answers is not in the document]
Comment (cwrau) [reply charts/t8s-cluster/templates/workload-cluster/cinder-csi-plugin.yaml:28]: I don't think the provided classes work for us; https://github.com/kubernetes/cloud-provider-openstack/blob/master/charts/cinder-csi-plugin/templates/storageclass.yaml

But what do you think?

File: charts/t8s-cluster/templates/workload-cluster/cinder-csi-plugin.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    secret:
+      enabled: true
+      name: cloud-config
```
[the remark this answers is not in the document]
Comment (cwrau) [reply charts/t8s-cluster/templates/workload-cluster/cinder-csi-plugin.yaml:31]: We do, that's why I'm setting the name to the existing secret 😁

File: charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  storageNamespace: kube-system
+  targetNamespace: kube-system
+  releaseName: ccm
+  valuesFrom:
+    - &valuesFrom
+      kind: Secret
+      name: {{ .Release.Name }}-ccm-cloud-config
+      optional: false
+      targetPath: cloudConfigContents
+      valuesKey: cloud.conf
```
[the remark this answers is not in the document]
Comment (cwrau) [reply charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml:32]: Yes, it needs to be deleted before it gets created again via the Helm chart, typical Helm-adoption problem

File: charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        name: {{ printf "%s-cloud-provider-openstack" .Release.Name }}
+        namespace: {{ .Release.Namespace }}
+      version: {{ printf "2.%d.x" (.Values.version.minor | int) }}
+  interval: 1h
+  kubeConfig:
+    secretRef:
+      name: {{ .Release.Name }}-kubeconfig
+  install:
+    remediation:
+      retries: -1
```
Comment (marvinWolff) [inline charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml:22]: I know we've used this a couple of times, but i never checked if the load is maybe rising through this.  
Does this have any impact on cpu usage?

Comment (cwrau) [reply charts/t8s-cluster/templates/workload-cluster/cloud-controller-manager.yaml:22]: _Maayybee_, but even if, only during the initial install.

And as we don't have a choice here 🤷

**Comment by marvinWolff** [comment]
I think this is good to go, what do you think @steutol ?

**Pull request merged**